File: leetcode/longestCommonPrefix/longestCommonPrefix.py
# Write a function to find the longest common prefix string amongst an array of strings.

# If there is no common prefix, return an empty string "".

# Example 1:
# Input: ["flower","flow","flight"]
# Output: "fl"

# Example 2:
# Input: ["dog","racecar","car"]
# Output: ""
# Explanation: There is no common prefix among the input strings.
# Note:

# All given inputs are in lowercase letters a-z.


class Solution:
    def longestCommonPrefix(self, strs):
        """
        :type strs: List[str]
        :rtype: str
        """
        rstr = ""
        strlist = list(zip(*strs))
        for prefix in strlist:
            c = list(set(prefix))
            if(len(c) == 1):
                rstr += c[0]
            else:
                break
        return rstr



if __name__ == "__main__":
    # strs = ["flower","flow","flight"]
    # strlist = list(zip(*strs))
    # print(strlist)
    solver = Solution()
    print(solver.longestCommonPrefix(["flower","flow","flight"]))
    print(solver.longestCommonPrefix(["dog","racecar","car"]))

File: leetcode/longestCommonPrefix/longestCommonPrefix.cpp
#include <iostream>
#include <string>
#include <vector>
#include <limits.h>
#include <math.h>
using namespace std;

class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {
        string result;
        int n = strs.size();
        if(n==1)
            return strs[0];
        int len = INT_MAX;
        for(int i=0; i<n; ++i){
            len = min(len, int(strs[i].size()));
        }
        // cout << len << endl;
        bool commonflag;
        for(int i=0; i< len; ++i){
            commonflag = false;
            for(int j=0; j< n-1; ++j)
            {
                if(strs[j][i] == strs[j+1][i])
                    commonflag = true; 
                else{
                    commonflag = false;
                    break;
                }
            }
            // cout << commonflag << endl;
            if(commonflag)
                result += strs[0].substr(i, 1);
            else 
                return result;   
                

        }
        return result;
        
    }
};


int main(){
    // for(int i=0; i<5; ++i){
    //     int j = 100;
    //     cout << i + j << endl;
    // }
    // if(~false)
    //     cout << "TURE" << endl;
    Solution solver;
    vector<string> strs{"flower","flow","flight"};
    cout << solver.longestCommonPrefix(strs) << endl;
    strs = {"a"};
    cout << solver.longestCommonPrefix(strs) << endl;
    return 0;
}